### PR TITLE
Support `guarded_by` attribute and related attributes inside C structs and support late parsing them

### LINKED ISSUE #20777: Thread Safety Analysis "guarded_by" attribute doesn't work for struct fields in C code
|  |  |
| --- | --- |
| Bugzilla Link | [20403](https://llvm.org/bz20403) |
| Version | trunk |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @AaronBallman,@Benvanik,@LebedevRI,@mikaelholmen,@nashidau,@zygoloid |

## Extended Description
```
$ cat demo.c
struct mutex {
        int x;
};

struct guarded_int {
        struct mutex mu;
        int value __attribute__((guarded_by(mu))); 
};

$ clang -c demo.c
demo.c:7:38: error: use of undeclared identifier 'mu'
        int value __attribute__((guarded_by(mu)));
                                            ^
1 error generated.
```

Clang compiles the code okay with "-x c++", so it seems to be a bug in how Thread Safety Analysis is wired into the C frontend.

### PR DESCRIPTION
**Pull request by pdherbemont**
This fixes #20777. This replaces #94216 which was reverted after being merged.

Previously the `guarded_by`, `pt_guarded_by`, `acquired_after`, and `acquired_before` attributes were only supported inside C++ classes or top level C/C++ declaration.

This patch allows these attributes to be added to struct members in C. These attributes have also now support experimental late parsing. This is off by default but can be enabled by passing `-fexperimental-late-parse-attributes`. This is useful for referring to a struct member after the annotated member. E.g.

```
struct Example {
  int a_value_defined_before __attribute__ ((guarded_by(a_mutex)));
  struct Mutex *a_mutex;
};
```

### CODE REVIEW

File: clang/lib/Parse/ParseDecl.cpp
Comment (hnrklssn) [inline clang/lib/Parse/ParseDecl.cpp:3368]: Why does C need to parse using `PotentiallyEvaluated` while C++ needs `Unevaluated`? What happens if we parse using `Unevaluted` in C?

Comment (pdherbemont) [reply clang/lib/Parse/ParseDecl.cpp]: I thought it wasn't working properly with `Unevaluated` but it seems like it actually does!

I have iterated on the patch further and simplified a bit by directly passing `EK_AttrArgument` to `EnterExpressionEvaluationContext` from within `ParseAttributeArgsCommon`. This allows me to remove `ParseGuardedAttribute()` and `ParseAcquiredAttribute()` and only rely on `ParseAttributeArgsCommon`.

This may have unforeseen side effects on some GNU/Clang/Microsoft attributes, but those are not caught by the test suite... So I am tempted to propose the simpler patch.

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (rapidsna) [inline clang/test/Sema/warn-thread-safety-analysis.c:6]: What's the reason to change it to take va_arg?

Comment (pdherbemont) [reply clang/test/Sema/warn-thread-safety-analysis.c:6]: To be able to test passing multiple arguments. On the previous iteration of the patch the attribute was parsed manually and we needed to make sure all cases where covered.

**Comment by pdherbemont** [comment]
I think this still needs review from @delcypher and @rapidsna

File: clang/docs/ThreadSafetyAnalysis.rst
Comment (aaronpuchert) [inline clang/docs/ThreadSafetyAnalysis.rst:772]: They are implemented, but still under `-Wthread-safety-beta` as far as I know, because there are some weaknesses. If you want to touch this here, I would suggest to write instead that support is still experimental.

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (aaronpuchert) [inline clang/test/Sema/warn-thread-safety-analysis.c:36]: I'm not entirely sure how nested `struct`s work in C, but should we be able to reference the parent from the child? At least in C++, you can declare objects of the child independent of the parent, so this wouldn't be meaningful.

This is an existing issue with with the `ACQUIRED_{AFTER,BEFORE}` implementation, but if you add a test for this (which is Ok to make sure it doesn't crash), you should probably also add a comment that this should perhaps be disallowed because it's not meaningful (as I understand).

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (aaronpuchert) [inline clang/test/Sema/warn-thread-safety-analysis.c:47]: Tricky, see #78131.

But as above, testing that it doesn't crash is fine.

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (pdherbemont) [reply clang/test/Sema/warn-thread-safety-analysis.c:36]: > I'm not entirely sure how nested structs work in C, but should we be able to reference the parent from the child?

I think so, it does read nicely it feels – I see why this could cause a problem though: if the child structure is used outside of the parent, there is an ambiguity. Maybe this should be restricted to anonymous struct? Or maybe we need to taint the child struct so that it's not reused outside of the parent?

Anyhow, I'll add a comment as suggested!

File: clang/docs/ThreadSafetyAnalysis.rst
Comment (pdherbemont) [reply clang/docs/ThreadSafetyAnalysis.rst]: Thanks! I documented those are "experimental".

**Review by rapidsna — APPROVED** [review]
LGTM. Let me know if you need me to merge this change.

**Comment by pdherbemont** [comment]
> LGTM. Let me know if you need me to merge this change.

Yes, please!

File: clang/test/Sema/warn-thread-safety-analysis.c
Comment (delcypher) [reply clang/test/Sema/warn-thread-safety-analysis.c:36]: @pdherbemont Did you add the comment? I can't find it in the diff shown on GitHub

**Review by delcypher — APPROVED** [review]
@pdherbemont Thanks for working on this LGTM. Please confirm that the comments that @aaronpuchert wanted have been added because I couldn't see them.

Once you've done that we can get this merged :)

**Comment by pdherbemont** [comment]
> @pdherbemont Thanks for working on this LGTM. Please confirm that the comments that @aaronpuchert wanted have been added because I couldn't see them.
> 
> Once you've done that we can get this merged :)

Done!

**Pull request merged**
